src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
             ImmutableArray<bool> memberIsReadOnly,
             ImmutableArray<NullableAnnotation> memberNullableAnnotations);
 
+        /// <summary>
+        /// Creates an <see cref="IMethodSymbol"/> whose <see cref="IMethodSymbol.MethodKind"/> is <see
+        /// cref="MethodKind.BuiltinOperator"/> for a binary operator. Built-in operators are commonly created for
+        /// symbols like <c>bool int.operator==(int v1, int v2)</c> which the language implicitly supports, even if such
+        /// a symbol is not explicitly defined for that type in either source or metadata.
+        /// </summary>
+        /// <param name="name">The binary operator name.  Should be one of the names from <see cref="WellKnownMemberNames"/>.</param>
+        /// <param name="returnType">The return type of the binary operator.</param>
+        /// <param name="leftType">The type of the left operand of the binary operator.</param>
+        /// <param name="rightType">The type of the right operand of the binary operator.</param>
+        /// <param name="isChecked">Whether or not this is a checked binary operator.  For example <c>int int.operator
+        /// checked +(int v1, int v2)</c>.</param>
+        /// <returns></returns>
+        public IMethodSymbol CreateBuiltinOperator(string name, ITypeSymbol returnType, ITypeSymbol leftType, ITypeSymbol rightType, bool isChecked)
+        {
+            // Can't check 'name' here as VB and C# support a different subset of names.
+
+            if (returnType is null)
+                throw new ArgumentNullException(nameof(returnType));
+
+            if (leftType is null)
+                throw new ArgumentNullException(nameof(leftType));
+
+            if (rightType is null)
+                throw new ArgumentNullException(nameof(rightType));
+
+            // When the types are dynamic, then virtually all operators are allowed and synthesized.
+            // So only check when neither are dynamic.
+            if (leftType.TypeKind != TypeKind.Dynamic && rightType.TypeKind != TypeKind.Dynamic)
+            {
+                CheckSupportedBinaryOperatorType(returnType, nameof(returnType));
+                CheckSupportedBinaryOperatorType(leftType, nameof(leftType));
+                CheckSupportedBinaryOperatorType(rightType, nameof(rightType));
+            }
+
+            return CommonCreateBuiltinOperator(name, returnType, leftType, rightType, isChecked);
+
+            void CheckSupportedBinaryOperatorType(ITypeSymbol type, string paramName)
+            {
+                // Enums have operators automatically synthesized for them.
+                if (type.TypeKind == TypeKind.Enum)
```

Comment:
>type.TypeKind == TypeKind.Enum [](http://example.com/codeflow?start=20&length=30)

Perhaps should let error types through. TypeKind won't be specific for them

Reply:
Ah.  Good case.  Will do that and add tests.